solver/monomial_solve.c

```diff
@@ -132,54 +138,58 @@ int solve_degenerate(spinor * const P, spinor * const Q, solver_params_t solver_
     }
     else{
       if(f==Qtm_pm_psi){   
-        iteration_count =  msolver_fp(P, Q, solver_params, max_iter, eps_sq, rel_prec, N, f, &Qtm_pm_psi_32);
-      }
-      else if(f==Q_pm_psi){     
-	iteration_count =  msolver_fp(P, Q, solver_params, max_iter, eps_sq, rel_prec, N, f, &Q_pm_psi_32);
+        iteration_count = msolver_fp(P, Q, solver_params, max_iter, eps_sq, rel_prec, N, f, &Qtm_pm_psi_32);
+      } else if(f==Q_pm_psi){     
+	iteration_count = msolver_fp(P, Q, solver_params, max_iter, eps_sq, rel_prec, N, f, &Q_pm_psi_32);
       } else if(f==Qsw_pm_psi){
         copy_32_sw_fields();
         iteration_count = msolver_fp(P, Q, solver_params, max_iter, eps_sq, rel_prec, N, f, &Qsw_pm_psi_32);
       } else {
         if(g_proc_id==0) printf("Warning: 32 bit matrix not available. Falling back to CG in 64 bit\n"); 
-        use_solver = CG;
+        solver_type = CG;
       }
     }
   } 
-  else if(use_solver == CG){
+  else if(solver_type == CG){
     iteration_count =  cg_her(P, Q, max_iter, eps_sq, rel_prec, N, f);
   }
-  else if(use_solver == BICGSTAB){
+  else if(solver_type == BICGSTAB){
      iteration_count =  bicgstab_complex(P, Q, max_iter, eps_sq, rel_prec, N, f);     
   }
 #ifdef DDalphaAMG 
-  else if (use_solver == MG)
+  else if (solver_type == MG)
     iteration_count =  MG_solver(P, Q, eps_sq, max_iter,rel_prec, N , g_gauge_field, f);
 #endif     
   else{
     if(g_proc_id==0) printf("Error: solver not allowed for degenerate solve. Aborting...\n");
     exit(2);
   }
-  if(g_debug_level > 0){
+
+  if(g_debug_level > 2){
     f(temp[0], P);
     diff(temp[0], temp[0], Q, VOLUME/2);
     double diffnorm = square_norm(temp[0], VOLUME/2, 1); 
     if( g_proc_id == 0 ){
       printf("# solve_degenerate residual check: %e\n", diffnorm);
     }
   }
-  if(g_debug_level > 0 || solver_params.external_inverter == QPHIX_INVERTER){
+  if(g_debug_level > 2 || solver_params.external_inverter == QPHIX_INVERTER){
     finalize_solver(temp, 1);
   }
+
   return(iteration_count);
 }
 
-int solve_mshift_oneflavour(spinor ** const P, spinor * const Q, solver_params_t* solver_params){
-  int iteration_count = 0;
+int solve_mms_tm(spinor ** const P, spinor * const Q,
+                 solver_params_t * solver_params){ 
+  int iteration_count = 0; 
+
+  // temporary field required by the QPhiX solve or by residual check
   spinor ** temp;
-  if(g_debug_level > 0 || solver_params->external_inverter == QPHIX_INVERTER){
+  if(g_debug_level > 2 || solver_params->external_inverter == QPHIX_INVERTER){
     init_solver_field(&temp, VOLUMEPLUSRAND/2, 1);
   }
-  
+
 #ifdef TM_USE_QPHIX
   if( solver_params->external_inverter == QPHIX_INVERTER ){
```

Comment:
Okay, I've tested that and it works (DDalphaAMG runs for a single iteration), but for some reason, even though I've specified that DDalphaAMG should do one of the shifts, all shifts are passed to QPhiX (rather than the remaining 9). Maybe I'm misunderstanding something...

Reply:
That's part of the tricks for having a good initial guess. Basically I'm running CGMMS for an higher shift, but I include also some of the smallest. The smallest will not be converged but it's on the way. The remaining part of the work is done by MG in really few iterations.

Reply:
@kostrzewa about this point I wanted to ask if in QphiX or its intereface there is a check regarding the shifts or convergence. Indeed CGMMS requires the shifts to be ordered from smallest to highest and here I'm not following this rule. Thus not all of them are converging but it's exactly what I want.
So if the CGMMS algorithm is "standard" then it should stop when the iterated shift is converged without checking if all of them did. Is it in this way? Is the second last shift non-converged in your case?